__tests__/integration/helpers/wallet.helper.js

```diff
@@ -0,0 +1,236 @@
+  return new Promise((resolve, reject) => {
+    hWallet.on('state', newState => {
+      if (newState === HathorWallet.READY) {
+        return resolve();
+      } if (newState === HathorWallet.ERROR) {
+        reject(new Error('Genesis wallet failed to start.'));
+      }
+    });
+  });
+}
+
+/**
+ * Translates the tx success event into a promise.
+ * Waits for the wallet event that indicates this transaction id has been fully integrated
+ * into the lib's local caches. Actions that depend on this state can be executed after the
+ * successful response from this function.
+ * @param {HathorWallet} hWallet
+ * @param {string} txId
+ * @param {number} [timeout] Timeout in milisseconds. Default value defined on test-constants.
+ * @returns {Promise<SendTxResponse>}
+ */
+export async function waitForTxReceived(hWallet, txId, timeout) {
+  const startTime = Date.now().valueOf();
+  let alreadyResponded = false;
+
+  // Only return the positive response after the transaction was received by the websocket
+  return new Promise(async (resolve, reject) => {
+    // Event listener
+    const handleNewTx = newTx => {
+      // Ignore this event if we didn't receive the transaction we expected.
+      if (newTx.tx_id !== txId) {
+        return;
+      }
+
+      // This is the correct transaction: resolving the promise.
+      resolveWithSuccess(newTx);
+    };
+    hWallet.on('new-tx', handleNewTx);
+
+    // Timeout handler
+    const timeoutPeriod = timeout || TX_TIMEOUT_DEFAULT;
+    setTimeout(async () => {
+      hWallet.removeListener('new-tx', handleNewTx);
+
+      // No need to respond if the event listener worked.
+      if (alreadyResponded) {
+        return;
+      }
+
+      /*
+       * If the timeout period passed and did not receive the new-tx event, probably the event
+       * was triggered even before this `waitForTxReceived` method was called.
+       * We'll try a last time to get the transaction data before rejecting this promise.
+       */
+      loggers.test.log(`Trying to solve timeout of ${timeoutPeriod}ms for ${txId}`);
+      const existingTx = hWallet.getTx(txId);
+      if (existingTx) {
+        return resolveWithSuccess(existingTx);
+      }
+
+      // Event listener did not receive the tx and it is not on local cache.
+      alreadyResponded = true;
+      reject(new Error(`Timeout of ${timeoutPeriod}ms without receiving tx ${txId}`));
+    }, timeoutPeriod);
+
+    async function resolveWithSuccess(newTx) {
+      const timeDiff = Date.now().valueOf() - startTime;
+      if (DEBUG_LOGGING) {
+        loggers.test.log(`Wait for ${txId} took ${timeDiff}ms.`);
+      }
+
+      if (alreadyResponded) {
+        return;
+      }
+      alreadyResponded = true;
+
+      /*
+       * Sometimes the tests happen too fast and it triggers timestamp protections on the fullnode.
+       * We will ensure there is some delay between transactions here.
+       */
+      await delay(50);
```

Comment:
From my experience with this `waitForTxReceived` I think the wallet methods were not prepared for such a rapid interaction as these integration tests make.

The multiple delays and re-confirmations I had to put in place to assure the transactions were really there were empirical and seem to be necessary for the data in memory to be correct every time. For instance, I tried removing this single `delay(50)` and more than 10 tests started failing each run.

I agree it would be good to improve this code, but I think this would require another level of investigation on how to best do so. For now, I tried keeping the delay times to a minimum to avoid impacting the overall test time and this specific comment was improved on [02c75bfe](https://github.com/HathorNetwork/hathor-explorer-service/pull/174/commits/02c75bfee477eaf1eb9761e28983bf26ce833efc).